### DISCUSSION TITLE
Is there a way to get the full page source ?

### DISCUSSION THREAD

**Discussion by ponienchen** [appium/appium#19044] [General]
driver.getPageSource() only gives me the current page in the viewPort but not the entire page including the contents outside the viewPort.

Is there way to do that ?

**Comment by jlipps** [comment]
The page source command returns all the elements known to Appium. Some platforms (like iOS) include elements outside the viewport. Some platforms (like Android) don't (because elements are dynamically constructed when the viewport scrolls; they don't exist otherwise). For Android, using something like the Espresso driver may give you a way to get list items that haven't been rendered yet, since Espresso runs in your app context and has access to internal app data.

**Reply by ponienchen** [reply]
> Some platforms (like iOS) include elements outside the viewport

But even so, for IOS, there is still not a way to get the page source for those elements outside the viewport, due to the limitations of appium ?

**Reply by jlipps** [reply]
iOS returns all elements, even those outside the viewport. At least that's how it's worked in my experience.

**Reply by KazuCocoa** [reply]
It might depend on the app's implementation. For example, table view/list view returned all elements even outside the screen, but grid view did not. (I might be wrong in details, but they have differences depending on the app's implementation)

**Reply by ponienchen** [reply]
Thanks for sharing your insights, Jonathan, Kazuaki.
Appreciated!
